src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/CPSProjectFactory.cs

```diff
@@ -0,0 +1,93 @@
+        public CPSProjectFactory(
+            VisualStudioProjectTracker projectTracker,
+            SVsServiceProvider serviceProvider,
+            VisualStudioWorkspaceImpl visualStudioWorkspace,
+            HostDiagnosticUpdateSource hostDiagnosticUpdateSource)
+        {
+            _projectTracker = projectTracker;
+            _serviceProvider = serviceProvider;
+            _visualStudioWorkspace = visualStudioWorkspace;
+            _hostDiagnosticUpdateSource = hostDiagnosticUpdateSource;
+            _externalErrorReporterMap = new Dictionary<string, IVsReportExternalErrors>(StringComparer.OrdinalIgnoreCase);
+        }
+
+        // internal for testing purposes only.
+        internal static CPSProject CreateCPSProject(VisualStudioProjectTracker projectTracker, IServiceProvider serviceProvider, IVsHierarchy hierarchy, string projectDisplayName, string projectFilePath, string language, Guid projectGuid, string projectTypeGuid, CommandLineArguments commandLineArguments)
+        {
+            return new CPSProject(commandLineArguments, projectTracker, reportExternalErrorCreatorOpt: null, hierarchy: hierarchy, language: language,
+                serviceProvider: serviceProvider, visualStudioWorkspaceOpt: null, hostDiagnosticUpdateSourceOpt: null, projectDisplayName: projectDisplayName,
+                projectFilePath: projectFilePath, projectGuid: projectGuid, projectTypeGuid: projectTypeGuid);
+        }
+
+        IProjectContext IProjectContextFactory.CreateProjectContext(
+            string languageName,
+            string projectDisplayName,
+            string projectFilePath,
+            Guid projectGuid,
+            string projectTypeGuid,
+            IVsHierarchy hierarchy,
+            CommandLineArguments commandLineArguments)
+        {
+            Contract.ThrowIfNull(hierarchy);
+
+            Func<ProjectId, IVsReportExternalErrors> getExternalErrorReporter = id => GetExternalErrorReporter(id, languageName);
+            return new CPSProject(commandLineArguments, _projectTracker, getExternalErrorReporter, projectDisplayName, projectFilePath,
+                projectGuid, projectTypeGuid, hierarchy, languageName, _serviceProvider, _visualStudioWorkspace, _hostDiagnosticUpdateSource);
+        }
+
+        private IVsReportExternalErrors GetExternalErrorReporter(ProjectId projectId, string languageName)
+        {
+            lock (_externalErrorReporterMap)
+            {
+                IVsReportExternalErrors errorReporter;
+                if (!_externalErrorReporterMap.TryGetValue(languageName, out errorReporter))
+                {
+                    string errorCodePrefix;
+                    if (!_projectLangaugeToErrorCodePrefixMap.TryGetValue(languageName, out errorCodePrefix))
+                    {
+                        Debug.Fail($"Unknown language '{languageName}'");
+                        _projectLangaugeToErrorCodePrefixMap.Add(languageName, languageName);
```

Comment:
@mavasani This line seems to be wrong. `_projectLangaugeToErrorCodePrefixMap` is immutable, so this line has no effect.